File: juan_vivi/infraestructura/lector.py
"""
Lector — infraestructura de entrada.
Lee archivos Excel AUTOFIN SP y los convierte en FilaEntrada del dominio.
Maneja limpieza básica de tipos (fechas como datetime, números, \xa0).
"""
import os
import openpyxl
from ..dominio.modelos import FilaEntrada


# ── Columnas esperadas (índice 0-based) ──────────────────────────
COL_NUMERO      = 0   # A — N°
COL_NOMBRE      = 1   # B — NOMBRE
COL_RUT         = 2   # C — RUT
COL_NOMBRE_PARA = 3   # D — NOMBRE PARA
COL_RUT_PARA    = 4   # E — RUT COMPRA PARA
COL_OPERACION   = 5   # F — OPERACIÓN
COL_PATENTE     = 6   # G — PATENTE
COL_TIPO        = 10  # K — COMUNIDAD | COMPRA PARA | None

COLUMNAS_REQUERIDAS = 11   # mínimo de columnas que debe tener la hoja


def _str_o_none(valor) -> str | None:
    """Convierte cualquier valor de celda a str limpio, o None si está vacío."""
    if valor is None:
        return None
    texto = str(valor).replace("\xa0", "").strip()
    return texto if texto else None


def _str_requerido(valor) -> str:
    """Como _str_o_none pero retorna '' en vez de None."""
    resultado = _str_o_none(valor)
    return resultado or ""


def leer_archivo(ruta: str) -> list:
    """
    Lee un Excel AUTOFIN SP y retorna lista[FilaEntrada].
    Lanza ValueError con mensaje descriptivo si el formato no es válido.
    """
    if not os.path.exists(ruta):
        raise FileNotFoundError(f"No se encontró el archivo: {ruta}")

    nombre_archivo = os.path.basename(ruta)

    try:
        wb = openpyxl.load_workbook(ruta, data_only=True)
    except Exception as e:
        raise ValueError(f"No se pudo abrir '{nombre_archivo}': {e}")

    ws = wb.active

    # Validar que tenga suficientes columnas
    if ws.max_column < COLUMNAS_REQUERIDAS:
        raise ValueError(
            f"'{nombre_archivo}' tiene solo {ws.max_column} columnas — "
            f"se esperan al menos {COLUMNAS_REQUERIDAS}."
        )

    filas: list[FilaEntrada] = []

    # La fila 1 es encabezado — empezar desde fila 2
    for num_fila, fila_excel in enumerate(ws.iter_rows(min_row=2, values_only=True), start=2):
        # Ignorar filas completamente vacías al final del archivo
        if all(c is None for c in fila_excel):
            continue

        entrada = FilaEntrada(
            numero=fila_excel[COL_NUMERO] if fila_excel[COL_NUMERO] is not None else 0,
            nombre=_str_requerido(fila_excel[COL_NOMBRE]),
            rut=_str_requerido(fila_excel[COL_RUT]),
            nombre_para=_str_o_none(fila_excel[COL_NOMBRE_PARA]),
            rut_para=_str_o_none(fila_excel[COL_RUT_PARA]),
            operacion=_str_o_none(fila_excel[COL_OPERACION]),
            patente=_str_o_none(fila_excel[COL_PATENTE]),
            tipo=_str_o_none(fila_excel[COL_TIPO]),
            num_fila_excel=num_fila,
        )
        filas.append(entrada)

    return filas


def leer_multiples(rutas: list) -> tuple[list, list]:
    """
    Lee varios archivos Excel y consolida las filas.
    Retorna (filas_consolidadas, errores_por_archivo).
    Los errores no detienen el proceso — se reportan y se sigue con el siguiente.
    """
    todas_las_filas: list[FilaEntrada] = []
    errores: list[str] = []

    for ruta in rutas:
        try:
            filas = leer_archivo(ruta)
            todas_las_filas.extend(filas)
        except (FileNotFoundError, ValueError) as e:
            errores.append(str(e))

    return todas_las_filas, errores

File: src-tauri/src/lector.rs
use calamine::{open_workbook_auto, Data, Reader};

use crate::modelos::FilaEntrada;

// ─── Índices de columna ──────────────────────────────────────────────────────
const COL_NUM: usize = 0;
const COL_NOMBRE: usize = 1;
const COL_RUT: usize = 2;
const COL_NOMBRE_PARA: usize = 3;
const COL_RUT_PARA: usize = 4;
const COL_OPERACION: usize = 5;
const COL_PATENTE: usize = 6;
const COL_TIPO: usize = 10;
const MIN_COLS: usize = 11;

// ─── Helpers ─────────────────────────────────────────────────────────────────

fn str_o_none(v: &Data) -> Option<String> {
    match v {
        Data::String(s) => {
            let t = s.trim().to_string();
            if t.is_empty() { None } else { Some(t) }
        }
        Data::Float(f) => Some(format!("{}", *f as i64)),
        Data::Int(i) => Some(i.to_string()),
        Data::Bool(b) => Some(b.to_string()),
        _ => None,
    }
}

fn str_req(v: &Data) -> String {
    str_o_none(v).unwrap_or_default()
}

fn es_fila_vacia(fila: &[Data]) -> bool {
    fila.iter().all(|c| matches!(c, Data::Empty))
}

// ─── Lectura ─────────────────────────────────────────────────────────────────

pub fn leer_archivo(ruta: &str) -> anyhow::Result<Vec<FilaEntrada>> {
    let mut wb = open_workbook_auto(ruta)?;
    let rango = wb
        .worksheet_range_at(0)
        .ok_or_else(|| anyhow::anyhow!("El archivo no tiene hojas"))??;

    if rango.width() < MIN_COLS {
        anyhow::bail!(
            "El archivo tiene {} columnas; se requieren al menos {}",
            rango.width(),
            MIN_COLS
        );
    }

    let mut filas: Vec<FilaEntrada> = Vec::new();

    for (idx, fila) in rango.rows().enumerate().skip(1) {
        if es_fila_vacia(fila) {
            continue;
        }

        let numero = match fila.get(COL_NUM) {
            Some(Data::Float(f)) => *f as i64,
            Some(Data::Int(i)) => *i,
            _ => 0,
        };

        let nombre = fila.get(COL_NOMBRE).map(str_req).unwrap_or_default();
        let rut = fila.get(COL_RUT).map(str_req).unwrap_or_default();
        let nombre_para = fila.get(COL_NOMBRE_PARA).and_then(str_o_none);
        let rut_para = fila.get(COL_RUT_PARA).map(str_req).unwrap_or_default();
        let operacion = fila.get(COL_OPERACION).map(str_req).unwrap_or_default();
        let patente = fila.get(COL_PATENTE).map(str_req).unwrap_or_default();
        let tipo = fila.get(COL_TIPO).map(str_req).unwrap_or_default();

        filas.push(FilaEntrada {
            numero,
            nombre,
            rut,
            nombre_para,
            rut_para,
            operacion,
            patente,
            tipo,
            num_fila_excel: idx + 1, // Excel es 1-based, idx es 0-based
        });
    }

    Ok(filas)
}

pub fn leer_multiples(rutas: &[String]) -> (Vec<FilaEntrada>, Vec<String>) {
    let mut todas: Vec<FilaEntrada> = Vec::new();
    let mut errores: Vec<String> = Vec::new();

    for ruta in rutas {
        match leer_archivo(ruta) {
            Ok(filas) => todas.extend(filas),
            Err(e) => errores.push(format!("{}: {}", ruta, e)),
        }
    }

    (todas, errores)
}
